### ISSUE TITLE
[Bug] Error when back-propagating through matmul with large dimension

### ISSUE THREAD

**Issue by Balandat** [getkeops/keops#30]
When back-propagating through operations that involve `matmul` with large matrix dimension, I run into the following error: 
```
RuntimeError: [KeOps] Arg number 6 : is not contiguous. Please provide 'contiguous' data array, as KeOps does not support strides. If you're getting this error in the 'backward' pass of a code using torch.sum() on the output of a KeOps routine, you should consider replacing 'a.sum()' with '(1. * a).sum()' or 'torch.dot(a.view(-1), torch.ones_like(a).view(-1))'. 
```

This happens at exactly dim 80, so I was able to trace this back to the special casing here: https://github.com/getkeops/keops/blob/26b4829727751acbee2378e3a48e247523e72e9a/pykeops/common/lazy_tensor.py#L1634-L1651

If I comment out that block and use the else for everything I don't observe this issue, so there must be something problematic going on there.

I don't have an stripped down repro, but the following, using gpytorch, is pretty concise:
[keops_backward_issue.ipynb.txt](https://github.com/getkeops/keops/files/3704847/keops_backward_issue.ipynb.txt)

cc @jacobrgardner, @gpleiss

**Comment by imagirom** [comment]
I am not sure whether this applies to your problem, but I was able to fix errors of gradients not being contiguous in the backward with a custom pytorch function:
```python
class ContiguousBackward(torch.autograd.Function):
    """
    Function to ensure contiguous gradient in backward pass. To be applied after PyKeOps reduction.
    """
    @staticmethod
    def forward(ctx, input):
        return input

    @staticmethod
    def backward(ctx, grad_output):
        return grad_output.contiguous()
```
I apply that function directly after the PyKeOps reduction:
```
tensor = lazy_tensor.sum(1)
tensor = ContiguousBackward().apply(tensor)
```
This is obviously not a proper solution (If it is it should probably be documented somewhere), but maybe it helps in your case too.

**Comment by joanglaunes** [comment]
Hello @Balandat ,
I removed this block of code so it should be ok now. 
To explain a little bit : this special implementation for N>80 is a trick we found this summer because @jacobgardner remarked that there was a huge decrease of performances of matmul with large dimensions. The trick did speed up computations a lot, starting around dims more or less greater than 80, so we decided to integrate it. Unfortunately we did not test the backprop and I am sorry it was buggy.
However later on we found for different test cases another very simple improvement by changing a precompiler directive for loop unrolling, and tests I ran last night show that in fact this precompiler directive was the real cause of the slowdown. So this trick for N>80 is no longer needed anyway.
Finally, thanks @imagirom for your comment, because this problem with contiguous arrays is indeed more general and recurrent, so we could maybe just include this "contiguous()" directive inside the backward (and even forward maybe) passes of KeOps reductions ; we will think about it.

Joan
